Adds the an express middleware that performs SSR

andreban opened the pull request.

andreban commented on README.md:
```diff
@@ -21,6 +21,7 @@ limitations under the License.
 A collection of AMP tools making it easier to publish and host AMP pages. The following tools are part of this project:
 
 - **[amp-ssr](./ssr):** a node library implementing server-side-rendering for AMP pages.
+- **[amp-ssr-express](./ssr-express)** an [express](http://expressjs.com/) middleware that transforms AMP pages on the fly.
```
done.

andreban commented:
Fixed the history.

sebastianbenz commented on ssr-express/README.md:
````diff
@@ -0,0 +1,131 @@
+
+The example bellow will transform HTML being loaded by express-static:
+```javascript
+const transformMiddleware = createTransformMiddleware(ampSSR);
````
suggestion: use ampSsrMiddleware

sebastianbenz replied:
suggestion: s/createTransformMiddleware/createAmpSsrMiddleware

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
````diff
@@ -0,0 +1,131 @@
+AMPs. eg: `https://example.com/page.html` becomes `https://example.com/page.html?amp`.
+
+```javascript
+// skipTransform receives a `string` containing an url as parameter, and returns a `boolean`.
````
s/url/URL

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,131 @@
+  return parsedUrl.searchParams.has('amp');
+};
+
+// ampUrl receives a `string` containing an url as parameter, and returns a `string` indicating
```
s/url/URL

sebastianbenz replied:
...there are more occurrences of `url`

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,131 @@
+
+// ampUrl receives a `string` containing an url as parameter, and returns a `string` indicating
+// the equivalent AMP url.
+const ampUrl = url => {
```
s/ampUrl/getAmpUrl

andreban replied:
done.

sebastianbenz commented on ssr-express/index.js:
```diff
@@ -0,0 +1,98 @@
+ *           returns the equivalent AMP url.
+ *
+ * @param {*} ampSSR the ampSSR instance to be used by this transformer.
+ * @param {*} options an optional object containing custom configureations for
```
configureations?

sebastianbenz replied:
Can you provide a type instead of `*`?

sebastianbenz replied:
Can you explicitly comment the options? I did it like this for runtime-version: https://github.com/ampproject/amp-toolbox/blob/master/runtime-version/lib/RuntimeVersion.js#L38

andreban replied:
done.

sebastianbenz commented on ssr-express/package.json:
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "amp-toolbox-ssr-express",
+  "version": "0.0.3",
+  "description": "Express middleware to for amp-toolbox-ssr",
```
s/to for/for/

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,131 @@
+## Introduction
+
+amp-ssr-express is an [express](http://expressjs.com/) middleware that applies AMP server-side-rendering on the fly, using the [amp-ssr](../ssr) component.
+
```
Can you explain first what it does and how you typically would use it?

andreban replied:
done.

ithinkihaveacat commented on ssr-express/README.md:
```diff
@@ -0,0 +1,131 @@
+limitations under the License.
+-->
+
+## Introduction
```
Is this only really useful to AMP Canonical sites? (Maybe say this?)

What happens if an AMP input file is not valid AMP?

andreban replied:
The first point was added on the latest commit. 

Re: the second point, the amp-ssr component doesn't perform validations, so invalid AMPs are transformed. Non-AMP pages, without ⚡, are not transformed. Those are more details of the amp-ssr component, than the middleware. We should add this explanation there and link it from this doc.

ithinkihaveacat commented on ssr-express/README.md:
```diff
@@ -0,0 +1,131 @@
+For the middleware to work, a route that renders AMP pages must be available. It will then rewrite the
+request to that route and apply server-side-rendering to the result.
+
+As an example, a website that has an SSR page at `https://example.com/page.html` would have it's AMP
```
I don't quite follow this paragraph. What does it mean for `/page.html` to be
rewritten to `/amp/page.html`?

What I was expecting was something like the below.

Before:

| URL | |
| --- | --- |
| `/page.html` | Canonical AMP |
| `/amp/page.html` | Does not exist |

After:

| URL | |
| --- | --- |
| `/page.html` | SSR'd AMP as transformed by `amp-ssr` (not valid AMP), contains `<link rel=amphtml href=/amp/page.html>` |
| `/amp/page.html` | Valid AMP, almost exactly the same as "before" version of `/page.html`, except it contains `<link rel=canonical href=/page.html>` and relative paths are rewritten. |

andreban replied:
Hum. This is not how it is working at the moment, but I think this is a better approach than the current one. @sebastianbenz WDYT?

ithinkihaveacat replied:
The adding/removing directories to fix URLs also seems a bit risky--even in AMP I'm not sure that you will be able to correctly rewrite all e.g. `../img/foo.png` to `../amp/img/foo.png` or whatever. (AMP version maybe better as `page.amp.html`, so that relative URLs don't need to change.)

sebastianbenz replied:
Good point Michael, we definitely should avoid re-writing the path!

ithinkihaveacat commented on ssr-express/README.md:
```diff
@@ -0,0 +1,131 @@
+const createAmpSsrMiddleware = require('amp-toolbox-ssr-express');
+const ampSsr = require('amp-toolbox-ssr');
+
+ampSsr.setConfig({
```
Is this config global, or local to each middleware instance? (If you want two for some reason.)

andreban replied:
amp-toolbox-ssr exports the following:
`module.exports = new DomTransformer(defaultConfig);`

So, every time it is required, a new DomTransformer object is created.

ithinkihaveacat replied:
Are you sure about that? `require(.)` is cached: "[every call to require('foo') will get exactly the same object returned](https://nodejs.org/docs/latest/api/modules.html#modules_caching))."

andreban replied:
Hum. Didn't know that. We may want to change `amp-ssr` if we want to have multiple instances then.

sebastianbenz replied:
+1

ithinkihaveacat commented on ssr-express/README.md:
````diff
@@ -0,0 +1,131 @@
+// This allows us to replace the parts needed before static handles the request.
+app.use(transformMiddleware);
+```
+## Cache server-side-rendered AMPs
````
Maybe call this out as a best practice/optional optimization: "Best Practice: Cache server-side-rendered AMPs".

andreban replied:
I'll rename to Best Practice.

andreban commented:
Please, PTAL

I've updated the behaviour of the middleware, as Michael suggested and switched to using an `amp`query parameter to identify AMP pages.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,76 @@
+## Introduction
+
+The amp-ssr-express is an [express](http://expressjs.com/) middleware that is typically used by
+developers who want to create a dynamic AMP website and take advantage of the performance boost
```
Why 'dynamic' AMP website? We also can't say typically if there are no users.

Suggestion for the introduction: "amp-ssr-express is an [express](http://expressjs.com/) middleware that optimizes page load times for websites using AMP for their canonical pages. The middleware uses the same [server-side-rendering](../srr) optimizations as the [Google AMP Cache](https://developers.google.com/amp/cache/).

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,76 @@
+As the server-side-rendered version of the content is ceases to be valid AMP, the component also
+handles serving the original content on an alternative URL to which server-side-rendering
+transformations are not applied. This alternative URL is used as the amphtml link of the
+transformed version.
```
I find the first two paragraphs quite hard to understand. Can you maybe explain this in a simpler way?

andreban replied:
updated.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,76 @@
+
+A valid AMP page is served on `https://example.com/index.html`.
+
+When the amp-ssr-express middleware is used, that URL will serve the server-side-rendered of the
```
s/server-side-rendered/server-side-rendered version

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,76 @@
+When the amp-ssr-express middleware is used, that URL will serve the server-side-rendered of the
+content.
+
+The original, valid AMP will then become available at `https://example.com/index.html?amp=`, and
```
`?amp=` looks a bit ugly. Maybe only `?amp`

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,76 @@
+content.
+
+The original, valid AMP will then become available at `https://example.com/index.html?amp=`, and
+amphtml link will be added to the server-side-rendered version at `https://example.com/index.html`.
```
Can you add more detail regarding `amphtml link`.

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,76 @@
+
+## Usage
+
+Usage follows the standard for most Express middlewares. It is important that the middleware is
```
suggestion: 'The AMP SSR middleware can be used like any other express middleware.'

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,76 @@
+To achieve best performance, those transformations shouldn't be applied for
+every request. Instead, transformations should only be applied the *first time*
+a page is requested, and the results then cached. Caching can happen on the CDN
+level, on the site's internal infrastructure (eg: Memcached), or even on the
```
s/eg/e.g.

andreban replied:
done.

sebastianbenz commented on ssr-express/lib/AmpSsrUrlScheme.js:
```diff
@@ -0,0 +1,72 @@
+  isAmpUrl(url) {
+    // http://example.com is used as the second parameter,
+    // as the URL constructor requires a valid domain.
+    const parsedUrl = new URL(url, 'https://example.com');
```
Extract the URL extraction into a method to encapsulate the ugly example.com hack. Personally, I'd prefer to simply use URLSearchParams directly. Something along the lines of:

```
function parseQueryParams(path) {
  const segments = path.split('?');
  if (segments.length === 1) {
    return new URLSearchParams('');
  }
  return new URLSearchParams(segments[1]);
}
```

andreban replied:
Changed to use URL.parse + URLSearchParams

sebastianbenz commented on ssr-express/lib/AmpSsrUrlScheme.js:
```diff
@@ -0,0 +1,72 @@
+   * Given a canonical URL, transforms it to the AMP equivalent.
+   *
+   * @param {string} canonicalUrl the canonical URL
+   * @returns {string} the equivalent AMP url for that page.
```
nit: s/url/URL

sebastianbenz replied:
Is this an URL or an path?

andreban replied:
`request.url` is what is passed to this method, and it send path + querystring. Looks more like a relative URL, as a path wouldn't contain the query string.

sebastianbenz commented on ssr-express/lib/AmpSsrUrlScheme.js:
```diff
@@ -0,0 +1,72 @@
+  /**
+   * Given an AMP URL, returns the Canonical equivalent
+   *
+   * @param {string} ampUrl the AMP URL
```
Is this an URL or an path?

sebastianbenz commented on ssr-express/lib/AmpSsrUrlScheme.js:
```diff
@@ -0,0 +1,72 @@
+ * When applying server-side-render AMPs, the resulting pages need to be linked to
+ * AMP equivalents.
+ */
+class AmpSsrUrlScheme {
```
Suggestion: rename to UrlMapping (as it has nothing to do with SSR or AMP per se)

andreban replied:
done.

sebastianbenz commented on ssr-express/lib/AmpSsrUrlScheme.js:
```diff
@@ -0,0 +1,72 @@
+const {URL} = require('whatwg-url');
+
+/**
+ * When applying server-side-render AMPs, the resulting pages need to be linked to
```
The documentation should be about what the class does, not what the greater context is. This class maps between two URLs.

andreban replied:
changed

sebastianbenz commented on ssr-express/lib/createAmpSsrMiddleware.js:
```diff
@@ -0,0 +1,108 @@
+ * @param {AmpSsrUrlScheme} options.ampSsrUrlScheme The scheme to be used when checking
+ * for AMP pages, rewriting to Canonical and generating amphtml links.
+ */
+const createAmpSsrMiddleware = (ampSsr, options) => {
```
How about creating a builder instead that provides a good default:

```
class AmpMiddleware {

  constructor() {
     this.ampSSR_ = ampSSR;
     this.options_ = {};
  }

  withAmpSsr(ampSsr) {
    this.ampSSR_ = ampSSR;
  }

  withOptions(options) {
    this.options_ = options;
  }

  create() {
    return (req, res, next) => {
       ...
    }
   }
}

module.exports = new AmpMiddleware();
```

You can then use it like:

```
const AmpMiddleware = require('amp-toolbox-ssr-express');

AmpMiddleware.create()

AmpMiddleware.withAmpSSR(...).create()
```

sebastianbenz replied:
We'll will also need an option for using versioned URLs.

andreban replied:
I like the builder, but most express middlewares (static, compression, bodyparser, etc) pass the configuration as an option when instantiating. I think it's worth keeping the API familiar to what other middlewares do.

sebastianbenz commented on ssr-express/lib/createAmpSsrMiddleware.js:
```diff
@@ -0,0 +1,108 @@
+      return;
+    }
+
+    // This is a request for the Canonical URL. Setup the middelware to transform the
```
nit/Canonical/canonical

andreban replied:
done.

sebastianbenz commented on ssr-express/lib/createAmpSsrMiddleware.js:
```diff
@@ -0,0 +1,108 @@
+
+      const body = Buffer.concat(chunks).toString('utf8');
+
+      // This is a request for the Canonical URL. Generate the AMP equivalent
```
s/Canonical/canonical

andreban replied:
done.

sebastianbenz commented on ssr-express/spec/lib/AmpSsrUrlSchemeSpec.js:
```diff
@@ -0,0 +1,47 @@
+describe('AmpSsrUrlScheme', () => {
+  const ampSsrUrlScheme = new AmpSsrUrlScheme('amp');
+
+  describe('isAmpUrl', () => {
```
Can you add a test for `ampSsrUrlScheme.isAmpUrl('/page?amp');`

andreban replied:
Added

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,81 @@
+transformations are not applied to this URL, and the original valid AMP is served.
+
+A amphtml link tag is added to the server-side-rendered version, linking it to the original valid
+AMP version hosted on the alernative URL.
```
s/alernative/alternative

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,81 @@
+The original, valid AMP will then become available at `https://example.com/index.html?amp`.
+
+An amphtml link will be added to the server-side-rendered version:
+`<link rel="amphtml" href="https://example.com/index.html?amp">`.
```
suggestion use a code block with '''

andreban replied:
done.

sebastianbenz commented on ssr-express/README.md:
```diff
@@ -0,0 +1,81 @@
+const path = require('path');
+const app = express();
+const createAmpSsrMiddleware = require('amp-toolbox-ssr-express');
+const ampSsr = require('amp-toolbox-ssr');
```
we shouldn't require to pass an amps instance

andreban replied:
done.

sebastianbenz commented on ssr-express/lib/UrlMapping.js:
```diff
@@ -0,0 +1,71 @@
+   * @returns {boolean} true, if the url is an AMP url.
+   */
+  isAmpUrl(url) {
+    const parsedUrl = URL.parse(url);
```
extract URLSearchParams instantiation into a separate method to avoid repeating the same code three times

andreban replied:
done

sebastianbenz commented on ssr-express/lib/UrlMapping.js:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  /**
+   * Given an AMP URL, returns the Canonical equivalent
```
s/Canonical/canonical

andreban replied:
done.

sebastianbenz commented on ssr-express/lib/UrlMapping.js:
```diff
@@ -0,0 +1,71 @@
+   * Given an AMP URL, returns the Canonical equivalent
+   *
+   * @param {string} ampUrl the AMP URL
+   * @returns {string} the equivalent Canonical URL for the AMP page.
```
s/Canonical/canonical

andreban replied:
done.

sebastianbenz commented on ssr-express/lib/createAmpSsrMiddleware.js:
```diff
@@ -0,0 +1,107 @@
+/**
```
Can we rename this file to AmpSsrMiddleware?

andreban replied:
done.

sebastianbenz commented on ssr-express/lib/createAmpSsrMiddleware.js:
```diff
@@ -0,0 +1,107 @@
+ * @param {UrlMapping} options.ampSsrUrlScheme The scheme to be used when checking
+ * for AMP pages, rewriting to canonical and generating amphtml links.
+ */
+const createAmpSsrMiddleware = (ampSsr, options) => {
```
Move ampSsr into options and use an default instance

andreban replied:
done.

sebastianbenz commented on ssr-express/demo/app.js:
```diff
@@ -0,0 +1,38 @@
+app.use(ampSsrMiddleware);
+
+const staticMiddleware = express.static(path.join(__dirname, '/public'));
+app.use(staticMiddleware);
```
I'd like to simplify the API to:

```
app.use(ampSsrMiddleware.create());
```

and the advanced case:

```
app.use(ampSsrMiddleware.create({
   urlMapping: ..,
   ampSsr: ampSsr
}));
```

andreban replied:
Makes sense. done.

andreban commented on ssr-express/README.md:
```diff
@@ -0,0 +1,81 @@
+provides the original content on an alternative URL. Server-side-rendering
+transformations are not applied to this URL, and the original valid AMP is served.
+
+A amphtml link tag is added to the server-side-rendered version, linking it to the original valid
```
done.

sebastianbenz commented on ssr-express/lib/UrlMapping.js:
```diff
@@ -61,11 +60,19 @@ class UrlMapping {
     return URL.format(parsedUrl);
   }
+
+  createURLSearchParams_(parsedUrl) {
```
super nit: parseUrlSearchParams would be a better name

sebastianbenz commented on ssr-express/demo/package.json:
```diff
@@ -0,0 +1,30 @@
+  },
+  "homepage": "https://github.com/ampproject/amp-toolbox#readme",
+  "dependencies": {
+    "amp-toolbox-ssr": "0.0.4",
```
s/0.0.4/*

sebastianbenz commented on ssr-express/spec/lib/AmpSsrMiddlewareSpec.js:
```diff
@@ -19,8 +19,8 @@ const MockExpressResponse = require('mock-express-response');
-  async transformHtml(body, options) {
-    return 'transformed: ' + options.ampUrl;
+  transformHtml(body, options) {
+    return Promise.resolve('transformed: ' + options.ampUrl);
```
Transformers do not need to be async at all.

andreban replied:
This is actually replacing the DomTransformer, which returns a Promise (https://github.com/ampproject/amp-toolbox/blob/master/ssr/lib/DomTransformer.js#L39-L44), not the internal transformers.

sebastianbenz commented:
ah ok - my faulte

On Fri, Jan 19, 2018 at 7:14 PM André Cipriani Bandarra <
notifications@github.com> wrote:

> *@andreban* commented on this pull request.
> ------------------------------
>
> In ssr-express/spec/lib/AmpSsrMiddlewareSpec.js
> <https://github.com/ampproject/amp-toolbox/pull/12#discussion_r162693880>:
>
> > @@ -19,8 +19,8 @@ const MockExpressResponse = require('mock-express-response');
>  const AmpSsrMiddleware = require('../../lib/AmpSsrMiddleware');
>
>  class TestTransformer {
> -  async transformHtml(body, options) {
> -    return 'transformed: ' + options.ampUrl;
> +  transformHtml(body, options) {
> +    return Promise.resolve('transformed: ' + options.ampUrl);
>
> This is actually replacing the DomTransformer, which returns a Promise (
> https://github.com/ampproject/amp-toolbox/blob/master/ssr/lib/DomTransformer.js#L39-L44),
> not the internal transformers.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amp-toolbox/pull/12#discussion_r162693880>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXOOK47mK0PnoMGQS4WNuh-x-oocHKCks5tMNuGgaJpZM4RhCL8>
> .
>
